packages/react-devtools-shared/src/backend/fiber/renderer.js

```diff
@@ -131,6 +133,55 @@ import type {
 import type {Source} from 'react-devtools-shared/src/shared/types';
 import {getStackByFiberInDevAndProd} from './DevToolsFiberComponentStack';
 
+const FIBER_INSTANCE = 0;
+// const VIRTUAL_INSTANCE = 1;
+
+// This type represents a stateful instance of a Client Component i.e. a Fiber pair.
+// These instances also let us track stateful DevTools meta data like id and warnings.
+type FiberInstance = {
+  kind: 0,
+  id: number,
+  parent: null | DevToolsInstance, // virtual parent
+  componentStack: null | string,
+  errors: null | Map<string, number>, // error messages and count
+  warnings: null | Map<string, number>, // warning messages and count
+  data: Fiber, // one of a Fiber pair
+};
+
+function createFiberInstance(fiber: Fiber): FiberInstance {
+  return {
+    kind: 0,
+    id: getUID(),
+    parent: null,
+    componentStack: null,
+    errors: null,
+    warnings: null,
+    data: fiber,
+  };
+}
+
+// This type represents a stateful instance of a Server Component or a Component
+// that gets optimized away - e.g. call-through without creating a Fiber.
+// It's basically a virtual Fiber. This is not a semantic concept in React.
+// It only exists as a virtual concept to let the same Element in the DevTools
+// persist. To be selectable separately from all ReactComponentInfo and overtime.
+type VirtualInstance = {
+  kind: 1,
+  id: number,
+  parent: null | DevToolsInstance, // virtual parent
```

Comment:
This is needed to get back from a VirtualInstance back into the Fiber tree.

We could store both children - e.g. a linked list like `child` and `next` like Fibers do. However, given either parent or children we can reverse engineer our location by traversing it until we hit a Fiber and then reverse engineer either parent or children.

The cheapest to store is the parent.

However, this will make `findHostInstancesForElementID` slower to implement because we'll need to traverse siblings and filter out any siblings not relevant to this Virtual Instance. That operation is just used for one ElementID at a time though so it's not so bad. We could also alternatively implement this on the front-end. E.g. find all HostComponents that the front-end thinks are the children a Component and then pass those ids to the backend. Since the front-end already has this tree.